### PR TITLE
API: Add the `aiida.engine.processes.control` module

### LINKED ISSUE #5258: python API for managing the AiiDA daemon & processes
Much of the utilities for managing the AiiDA daemon & Processes are currently buried in the command line interface.

It would be nice to expose these through the python API.

This feature requests originates e.g. from
https://github.com/microsoft/aiida-dynamic-workflows/blob/main/aiida_dynamic_workflows/control.py

Mentioning @jbweston (in case of further input) and @muhrin (since he signalled interest).

I'll have a look into this later today if I find time

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5258 

This module abstracts functionality that was implemented directly in the
`aiida.cmdline.commands.cmd_process` module that allows to interact with
active processes. By extracting it to a separate module, this
functionality can now also be easily used by users from interfaces other
than the CLI, such as the shell, or a notebook.

### CODE REVIEW

**Comment by sphuber** [comment]
Docs build is failing but I haven't been able to fix it yet. The problem is coming from the type hints. Adding a rule to the ignore nitpicks didn't seem to help.

File: aiida/engine/processes/control.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+def play_processes(
+    processes: list[ProcessNode] | None = None,
+    all_entries: bool = False,
+    timeout: float = 5.0,
+    wait: bool = False
+) -> None:
+    """Play (unpause) paused processes.
+
+    .. note:: Requires the daemon to be running, or processes will be unresponsive.
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:44]: I guess adding a check for the daemon status in the function would be expensive?

File: aiida/engine/processes/control.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    :raises ``ProcessTimeoutException``: If the processes do not respond within the timeout.
+    """
+    if processes and all_entries:
+        raise ValueError('cannot specify processes when `all_entries = True`.')
+
+    if not processes and all_entries:
+        processes = get_active_processes()
+
+    if not processes:
+        raise ValueError('no processes specified and `all_entries = False`.')
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:59]: I don't think that's true, as `get_active_processes()` may also just have returned an empty list above.

Maybe easier to just say "no processes selected".

And I would consider this a valid selection, i.e. I would not raise an error but rather use something like `LOGGER.report` (or warning) and `return`

File: aiida/engine/processes/control.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    :param wait: Set to ``True`` to wait for process response, for ``False`` the action is fire-and-forget.
+    :raises ``ProcessTimeoutException``: If the processes do not respond within the timeout.
+    """
+    if processes and all_entries:
+        raise ValueError('cannot specify processes when `all_entries = True`.')
+
+    if not processes and all_entries:
+        processes = get_active_processes()
+
+    if not processes:
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:87]: see above

File: aiida/engine/processes/control.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    all_entries: bool = False,
+    timeout: float = 5.0,
+    wait: bool = False
+) -> None:
+    """Pause running processes.
+
+    .. note:: Requires the daemon to be running, or processes will be unresponsive.
+
+    :param processes: List of processes to play.
+    :param all_entries: Play all paused processes.
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:76]: Pause all playing processes?

File: aiida/engine/processes/control.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    controller = get_manager().get_process_controller()
+    _perform_actions(processes, controller.pause_process, 'pause', 'pausing', 'paused', wait, timeout)
+
+
+def kill_processes(processes: list[ProcessNode], timeout: float = 5.0, wait: bool = False) -> None:
+    """Kill running processes.
+
+    .. note:: Requires the daemon to be running, or processes will be unresponsive.
+
+    :param processes: List of processes to play.
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:99]: why no `all_entries` here?

Can certainly imagine someone wanting to kill all running processes

File: aiida/engine/processes/control.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+def _perform_actions(
+    processes: list[ProcessNode],
+    action: t.Callable,
+    infinitive: str,
+    present: str,
+    past: str,
+    wait: bool = False,
+    timeout: float = None
+) -> None:
+    """."""
```
Comment (ltalirz) [inline aiida/engine/processes/control.py:117]: let's expand a little on that

**Review by ltalirz — COMMENTED** [review]
This is great, thanks a lot @sphuber !

Splitting the move between files into a separate commit made the review much more straightforward.

I don't have any major comments, just a few suggestions

File: aiida/engine/processes/control.py (hunk 1 above)
Comment (sphuber) [reply aiida/engine/processes/control.py:44]: That's why I didn't add it, but actually from the API it is quite fast
```
In [6]: %timeit get_daemon_client().is_daemon_running
63.1 µs ± 1.15 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
So might as well add it.

File: aiida/engine/processes/control.py (hunk 5 above)
Comment (sphuber) [reply aiida/engine/processes/control.py:99]: Could be dangerous? It isn't there currently in the API, and we haven't really had anyone ask for it. But happy to add it if you think that's fine

**Comment by chrisjsewell** [comment]
looks good cheers 👍

**Review by ltalirz — APPROVED** [review]
Thanks for addressing the comments; good to go from my side as well.

**Pull request merged**